File: Algebras/FixedMRegion/fmr/RCurve.cpp
/* 
 * This file is part of libfmr
 * 
 * File:   RCurve.cpp
 * Author: Florian Heinz <fh@sysv.de>
 * 
 * Created on September 20, 2016, 4:22 PM

//paragraph [1] Title: [{\Large \bf \begin {center}] [\end {center}}]
//[TOC] [\tableofcontents]

[1] Implementation of the class ~RCurve~

[TOC]

1 Overview

The class ~RCurve~ describes curved line segments from which a ~RFace~ of
a ~CRegion~ is constructed.
The RCurve is defined by:

Curve type: ~type~
Startpoint: ~off~
Angle: ~angle~
Type-specific parameters: ~params~

Currently, three types are defined:

T : A trochoidal line segment
Parameters: <a> <b> <toff> <rotation>
Parametric functions:
fx(t) = a * t * rot - b * (sin(t * rot + toff) - sin(toff))
fy(t) = b * (cos(t * rot + toff) - cos(toff))

R : A ravdoidal line segment
Parameters: <hp> <cd> <toff> <rotation>
Parametric functions:
fx(t) = hp * (2 * t * rot - sin(2 * (t * rot + toff)) + sin(2 * toff)) +
        cd *               (cos     (t * rot + toff)  - cos(    toff))
fy(t) = hp * (cos(2 * (t * rot + toff)) - cos(2 * toff)) +
        cd * (sin     (t * rot + toff)  - sin(    toff))

S : A straight line segment
Parameters: <xd> <yd>
Parametric functions:
fx(t) = xd * t
fy(t) = yd * t

When implementing more types, it should be ensured that fx(0) = fy(0) = 0,
so that the point ~off~ is really the startpoint of the line segment.

A point on the curve at time ~t~ is then calculated as:
Point(fx(t), fy(t)).rotate(~angle~) + ~off~

*/

#include "fmr_RCurve.h"
#include "fmr_Trochoid.h"
#include "fmr_Ravdoid.h"
#include "fmr_SegT.h"
#include "fmr_ISSegCurve.h"

#define MIN(a,b,c,d) std::min(std::min(std::min(a,b),c),d)
#define MAX(a,b,c,d) std::max(std::max(std::max(a,b),c),d)

using namespace fmr;

/*
2 Construct from RList

Constructs an ~RCurve~ object from an RList representation.

*/
RCurve::RCurve(RList& r) {
    off = Point(r[0].getNr(), r[1].getNr());
    angle = r[2].getNr();
    type = r[3].getSym();
    for (int i = 4; i < r.size(); i++) {
        params.push_back(r[i].getNr());
    }
}

/*
3 ~getCurve~

Reconstructs the underlying curvetype (trochoid, ravdoid, segment)
from this RCurve. These are used for example for intersection checks then.

*/
Curve* RCurve::getCurve() {
    Curve *ret;
    
    if (type == "T") { // Construct a Trochoid
        double a = params[0];
        double b = params[1];
        double toff = params[2];
        double rot = params[3];
        double xoff = off.x - a*toff + b*sin(toff);
        double yoff = off.y + a - b*cos(toff);
        
        ret = new Trochoid(a, b, xoff, yoff, toff, rot);
    } else if (type == "R") { // Construct a Ravdoid
        double hp = params[0];
        double cd = params[1];
        double toff = params[2];
        double rot = params[3];
        double xoff = off.x - hp*(2*toff - sin(2*toff)) - cd * cos(toff);
        double yoff = off.y - hp*cos(2*toff) - cd * sin(toff);
        
        ret = new Ravdoid(hp, cd, xoff, yoff, toff, NAN, NAN, NAN, NAN, rot);
    } else if (type == "S") { // Construct a line segment
        ret = new SegT(Seg(Point(off.x, off.y),
                           Point(off.x+params[0], off.y+params[1])));
    } else
        return NULL;
    
    ret->setTransformation(off, angle); // Copy the stored transformation
    
    return ret;
}

/*
4 ~intersections~

Calculate all intersections of a line segment with this ~RCurve~ segment.

*/
std::vector<Point> RCurve::intersections (Seg s) {
    Curve *c = getCurve(); // First, construct the corresponding Curve
    if (!c) { // Can fail if the type is invalid/unsupported by getCurve()
        std::vector<Point> none;
        return none;
    }
    
    SegT segt(s.rotate(off, -angle)); // Correct the orientation of the segment
    ISSegCurve is(segt, *c);
    
    std::vector<Point> ret = is.findIntersection(); // Get intersections
    // Undo the transformation above
    for (int nrpoint = 0; nrpoint < ret.size(); nrpoint++) {
        ret[nrpoint] = ret[nrpoint].rotate(off, angle);
    }
    
    return ret;
}

/*
5 ~boundingBox~

Calculate the bounding box of this line segment. This is not necessarily a
minimal bounding box.

*/
BoundingBox RCurve::boundingBox() {
    // Check, if we have the boundingBox already cached
    if (bb.valid()) {
        return bb;
    }
    
    if (type == "T") {
        // Curve type is Trochoid, expand parameters
        double a = params[0];
        double b = params[1];
        double toff = params[2];
        double rot = params[3];
        double xoff = off.x - a*toff + b*sin(toff);
        double yoff = off.y + a - b*cos(toff);
        
        // Calculate maximum possible x and y values
        double x1 = xoff + a*toff - b;
        double x2 = xoff + a*(toff+rot) + b;
        double y1 = yoff - a - b;
        double y2 = yoff - a + b;
        // and create the resulting bounding box
        bb.update(Point(x1, y1));
        bb.update(Point(x2, y2));
    } else if (type == "R") {
        // Curve type is Ravdoid, expand parameters
        double hp = params[0];
        double cd = params[1];
        double toff = params[2];
        double rot = params[3];
        double xoff = off.x - hp*(2*toff - sin(2*toff)) - cd * cos(toff);
        double yoff = off.y - hp*cos(2*toff) - cd * sin(toff);
        
        // Calculate maximum possible x and y values
        double x1 = xoff + hp * (2* toff      - 1) - cd;
        double x2 = xoff + hp * (2*(toff+rot) + 1) + cd;
        double y1 = yoff - hp - cd;
        double y2 = yoff + hp + cd;
        // and create the resulting bounding box
        bb.update(Point(x1, y1));
        bb.update(Point(x2, y2));
    } else if (type == "S") {
        // Curve type is Straight segment, calculate start and end point
        Point start(off.x, off.y);
        Point end(off.x+params[0], off.y+params[1]);
        // and create the resulting bounding box
        bb.update(start);
        bb.update(end);
    }
    
    // The segment is possibly rotated, rotate the corners of the bounding box
    // and calculate the bounding box of that rotated box.
    bb.rotate(off, angle);
    
    return bb;
}

/*
6 ~toSegs~

Approximate this RCurve by ~nrsegs~ straight line segments.

*/
std::vector<Seg> RCurve::toSegs(int nrsegs) {
    std::vector<Seg> ret;
    
    if (type == "S") {
        // Special case: We already have a straight segment, just use that.
        ret.push_back(Seg(Point(off.x, off.y), 
                          Point(off.x+params[0], off.y+params[1])));
    } else {
        // for Trochoids and Ravdoids sample nrsegs+1 points and construct
        // segments from those.
        Curve *c = getCurve();
        Point prev;
        for (int i = 0; i < nrsegs+1; i++) {
            Point p = c->project(((double)i)/((double)(nrsegs+1)));
            if (prev.valid()) {
                ret.push_back(Seg(prev, p).rotate(c->t_center, c->t_angle));
            }
            prev = p;
        }
        delete c;
    }
    
    return ret;
}

/*
7 ~toRList~

Return the RList representation for this RCurve.

*/
RList RCurve::toRList() {
    RList ret;
    
    ret.append(off.x);
    ret.append(off.y);
    ret.append(angle);
    ret.appendSym(type);
    for (int i = 0; i < params.size(); i++) {
        ret.append(params[i]);
    }
    
    return ret;
}

File: Javagui/viewer/hoese/algebras/fixedmregion/RCurve.java
package viewer.hoese.algebras.fixedmregion;

import java.util.LinkedList;
import java.util.List;

/**
 * Class RCurve represents a curved line segment of a CFace/CRegion
 * 
 * Currently three types of line segment are defined:
 *
 * - Trochoid      (type 'T')
 * - Ravdoid       (type 'R')
 * - Straight line (type 'S')
 * 
 * The time interval of the parametric functions is always normalized to [0;1]
 * 
 * @author Florian Heinz <fh@sysv.de>
 */
enum SegmentType {
   Trochoid, Ravdoid, Straight, Unknown;

   @Override
   public String toString(){
      switch (this){
        case Trochoid: return "T";
        case Ravdoid : return "R";
        case Straight: return "S";
        default: return null;
      }
   }

   static SegmentType  fromString(String s){
      if(s.equals("T")) return Trochoid; 
      if(s.equals("R")) return Ravdoid; 
      if(s.equals("S")) return Straight; 
      return Unknown;
   }

};



public class RCurve {
    /** Start point of this RCurve */
    private Point start;
    /** Rotation angle (center is always the start point) */
    private double angle;
    /** Type of curve */
    private SegmentType type;
    /** curve-specific parameters */
    private double[] params;

    /**
     * Construct an rcurve from all parameters.
     * 
     * @param start  startpoint
     * @param angle  rotation angle
     * @param type   type of curve (T,R,S)
     * @param params curve specific parameters 
     */
    public RCurve(Point start, double angle, SegmentType type, double[] params) {
        this.start = start;
        this.type = type;
        this.params = params;
        this.angle = angle;
    }
    
    /**
     * Construct an rcurve from its parameters, but calculate the initial
     * rotation for a given center.
     * 
     * @param start  startpoint
     * @param type   type of curve (T,R,S)
     * @param params curve specific parameters 
     * @param center rotation center
     * @param angle  rotation angle
     */
    public RCurve(Point start, SegmentType type, double[] params, Point center, double angle) {
        this.start = start;
        this.type = type;
        this.params = params;
        rotate (angle, center);
    }
    
    /**
     * Set the rotation parameters for this RCurve from an angle and a given
     * center.
     * After rotation the start point, the new center will be the start point.
     * 
     * @param angle  the rotation angle
     * @param center the center point
     */
    public final void rotate (double angle, Point center) {
        this.angle = angle;
        start = start.rotate(center, angle);
    }
    
    /**
     * Calculate the point on the curve at time t for type "Straight" (S)
     * 
     * @param t the time
     * @return point on curve
     */
    public Point straight(double t) {
        double xd = params[0];
        double yd = params.length == 2 ? params[1] : 0;
        
        Point off = new Point(xd, yd).mul(t);
        
        return start.add(off);
    }

    /**
     * Calculate the point on the curve at time t for type "Trochoid" (T)
     * 
     * @param t the time
     * @return point on curve
     */
    public Point trochoid(double t) {
        double a = params[0];
        double b = params[1];
        double toff = params[2];
        double rot = params[3];

        double x = a * t * rot - b * (Math.sin(t * rot + toff) - Math.sin(toff));
        double y = b * (Math.cos(t * rot + toff) - Math.cos(toff));
        
        Point off = new Point(x, y);

        return start.add(off);
    }

    /**
     * Calculate the point on the curve at time t for type "Ravdoid" (R)
     * 
     * @param t the time
     * @return point on curve
     */
    public Point ravdoid(double t) {
        double hp = params[0];
        double cd = params[1];
        double toff = params[2];
        double rot = params[3];

        double x = hp * (2 * t * rot - Math.sin(2 * (t * rot + toff)) + Math.sin(2 * toff))
                + cd * (Math.cos(t * rot + toff) - Math.cos(toff));

        double y = hp * (Math.cos(2 * (t * rot + toff)) - Math.cos(2 * toff))
                + cd * (Math.sin(t * rot + toff) - Math.sin(toff));
        
        Point off = new Point(x, y);

        return start.add(off);
    }

    /**
     * Calculate the point on this curve for time t.
     * 
     * @param t the time
     * @return Point on curve
     */
    public Point f(double t) {
        Point ret = null; 
        switch (type) {
            case Straight:
                ret = straight(t);
                break;
            case Trochoid:
                ret = trochoid(t);
                break;
            case Ravdoid:
                ret = ravdoid(t);
                break;
        }

        return ret.rotate(start, angle);
    }

    /**
     * Calculate the x coordinate on this curve for time t.
     * 
     * @param t the time
     * @return x coordinate
     */
    public double fx(double t) {
        return f(t).x;
    }

    /**
     * Calculate the y coordinate on this curve for time t.
     * 
     * @param t the time
     * @return y coordinate
     */
    public double fy(double t) {
        return f(t).y;
    }
    
    /**
     * Calculate a list of nrpoints points of this curve to approximate its
     * shape.
     * The more points, the better the approximation.
     * 
     * @param nrpoints Number of points to calculate
     * @return list of points
     */
    public List<Point> getPoints (int nrpoints) {
        List<Point> ret = new LinkedList();
        if (type.equals("S"))
            nrpoints = 1;
        for (int i = 0; i < nrpoints; i++) {
            double t = ((double)i) / ((double)nrpoints);
            ret.add(f(t));
        }
        
        return ret;
    }
    
    /**
     * Construct an RCurve from its nested list representation.
     * 
     * @param nl Nested List representation
     * @return rcurve object
     */
    public static RCurve fromNL(NL nl) {
        double sx = nl.get(0).getNr();
        double sy = nl.get(1).getNr();
        double angle = nl.get(2).getNr();
        SegmentType type = SegmentType.fromString(nl.get(3).getSym());
        double[] params = null;
        if (nl.size() > 4) {
            params = new double[nl.size() - 4];
            for (int i = 4; i < nl.size(); i++) {
                params[i - 4] = nl.get(i).getNr();
            }
        }

        return new RCurve(new Point(sx, sy), angle, type, params);
    }
    
    /**
     * Convert this RCurve to its nested list representation
     * 
     * @return nested list representation of this rcurve
     */
    public NL toNL() {
        NL nl = new NL();
        nl.addNr(start.x);
        nl.addNr(start.y);
        nl.addNr(angle);
        nl.addSym(type.toString());
        for (double d : params) {
            nl.addNr(d);
        }
        
        return nl;
    }
    
    /**
     * Get a string representation of this RCurve
     * 
     * @return string representation
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("( ").append(start.x).append(" ")
                .append(start.y).append(" ").append(angle).append(" ").append(type);
        if (params != null) {
            for (double v : params) {
                sb.append(" ").append(v);
            }
        }
        sb.append(" )");
        return sb.toString();
    }

    /**
     * Get the angle of this RCurve segment
     * 
     * @return angle of segment
     */
    public double getAngle() {
        return angle;
    }

    /**
     * Set the angle of this RCurve segment
     * 
     * @param angle angle of segment
     */
    public void setAngle(double angle) {
        this.angle = angle;
    }
}
